pkg/controller/dataimportcron-format-handlers.go

```diff
@@ -0,0 +1,141 @@
+)
+
+var formatHandlers = map[cdiv1.DataImportCronSourceFormat]formatHandler{
+	cdiv1.DataImportCronSourceFormatSnapshot: snapshotFormatHandler{},
+	cdiv1.DataImportCronSourceFormatPvc:      pvcFormatHandler{},
+}
+
+type formatHandler interface {
+	populateDataSourceSpec(*cdiv1.DataSourceSpec, *cdiv1.DataVolumeSourcePVC)
+	handleSourceCreation(context.Context, *DataImportCronReconciler, *cdiv1.DataImportCron, *cdiv1.DataVolume, *storagev1.StorageClass) error
+	updateOnSuccess(context.Context, *DataImportCronReconciler, *cdiv1.DataImportCron, client.Object)
+}
+
+type snapshotFormatHandler struct{}
+
+func (snapshotFormatHandler) populateDataSourceSpec(dataSourceSpec *cdiv1.DataSourceSpec, sourcePVC *cdiv1.DataVolumeSourcePVC) {
+	dataSourceSpec.Source = cdiv1.DataSourceSource{
+		Snapshot: &cdiv1.DataVolumeSourceSnapshot{
+			Namespace: sourcePVC.Namespace,
+			Name:      sourcePVC.Name,
+		},
+	}
+}
+
+func (snapshotFormatHandler) handleSourceCreation(ctx context.Context, r *DataImportCronReconciler, dataImportCron *cdiv1.DataImportCron, dataVolume *cdiv1.DataVolume, dvStorageClass *storagev1.StorageClass) error {
+	dataSourceName := dataImportCron.Spec.ManagedDataSource
+	digest := dataImportCron.Annotations[AnnSourceDesiredDigest]
+	if digest == "" {
+		return nil
+	}
+	dvName, err := createDvName(dataSourceName, digest)
+	if err != nil {
+		return err
+	}
+
+	className, err := cc.GetSnapshotClassForSmartClone(dataVolume.Name, &dvStorageClass.Name, r.log, r.client)
+	if err != nil {
+		return err
+	}
+	labels := map[string]string{
+		common.CDILabelKey:       common.CDILabelValue,
+		common.CDIComponentLabel: "",
+	}
+	desiredSnapshot := &snapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dvName,
+			Namespace: dataImportCron.Namespace,
+			Labels:    labels,
+		},
+		Spec: snapshotv1.VolumeSnapshotSpec{
+			Source: snapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &dvName,
+			},
+			VolumeSnapshotClassName: &className,
+		},
+	}
+	r.setDataImportCronResourceLabels(dataImportCron, desiredSnapshot)
+
+	currentSnapshot := &snapshotv1.VolumeSnapshot{}
+	if err := r.client.Get(context.TODO(), client.ObjectKeyFromObject(desiredSnapshot), currentSnapshot); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		cc.AddAnnotation(desiredSnapshot, AnnLastUseTime, time.Now().UTC().Format(time.RFC3339Nano))
+		if err := r.client.Create(ctx, desiredSnapshot); err != nil {
+			return err
+		}
+		return nil
+	}
+	if cc.IsSnapshotReady(currentSnapshot) {
+		// Clean up DV/PVC as they are not needed anymore
+		r.log.Info("Deleting dv/pvc as snapshot is ready", "name", desiredSnapshot.Name)
+		if err := r.deleteDvPvc(ctx, desiredSnapshot.Name, desiredSnapshot.Namespace); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (snapshotFormatHandler) updateOnSuccess(ctx context.Context, r *DataImportCronReconciler, dataImportCron *cdiv1.DataImportCron, sourceObj client.Object) {
```

Comment:
Any reason we are passing the context and reconciler in and then not using it? I am guessing it is a leftover.

Reply:
yep